### PR TITLE
CAD-1344 CAD-1384:  `cardano-tx-generator` update for 1.16

### PR DESCRIPTION
**Pull request by deepfire**
- update for compat with `ouroboros-network` `master` & `cardano-api` & node 1.16 -- we depended on custom branches before.
  - this mainly was about the switch to HFC types and the typed implementation of `cardano-api`
- implement extra Tx payload for Shelley
- polymorphic code replaces huge chunks of duplicated code, that were slightly different between Byron and Shelley
- no longer using `cardano-shell`
- factor the `Benchmark` type and module as specification for benchmark
- factor the `Era` type and module as non-benchmark, era-specific parameters
- implement the partial options monoid for the benchmark specification
- factor the `Genesis` module to handle genesis fund extraction
- factor the `Tx` module for transaction creation functionality

### CODE REVIEW

File: cardano-tx-generator/src/Cardano/Benchmarking/GeneratorTx/Genesis.hs [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+-- Shelley-specific imports
+import qualified Ouroboros.Consensus.Shelley.Ledger.Ledger as Shelley
+
+import Cardano.Api.Typed
+import Cardano.Benchmarking.GeneratorTx.Era
+import Cardano.Benchmarking.GeneratorTx.Tx
+import Cardano.Benchmarking.GeneratorTx.Tx.Byron
+
+
+keyAddress :: Era era -> SigningKeyOf era -> Address era
```
Comment (CodiePP) [inline cardano-tx-generator/src/Cardano/Benchmarking/GeneratorTx/Genesis.hs:47]: would it make sense to define a class with instances for Byron, Shelley that provide these functions?

**Review by CodiePP — CHANGES_REQUESTED** [review]
:+1: 
only: please, squash the commits into one

**Comment by deepfire** [comment]
> only: please, squash the commits into one

Done.

File: cardano-tx-generator/src/Cardano/Benchmarking/GeneratorTx/Genesis.hs (hunk 1 above)
Comment (deepfire) [reply cardano-tx-generator/src/Cardano/Benchmarking/GeneratorTx/Genesis.hs:47]: The GADT dispatch is based on the `cardano-api` model actually, which mostly isn't class-based in the era aspect.

So if we'd choose a different model, we'd have this conflict to paper over.

Also, it's not clear what the class-based approach would help us with..

**Review by CodiePP — APPROVED** [review]

**Pull request merged**